kappa.js

```diff
@@ -0,0 +1,319 @@
+          } else ondone()
+        })
+      } else ondone()
+    }
+
+    function ondone () {
+      self._opened = true
+      self._run()
+      cb()
+    }
+  }
+
+  close (cb) {
+    const self = this
+    this.pause()
+    this._closing = true
+    if (this.status === Status.Running) return this.once('ready', close)
+    else close()
+    function close () {
+      let pending = 1
+      if (self._source.close) ++pending && self._source.close(cb)
+      if (self._view.close) ++pending && self._view.close(cb)
+      done()
+      function done () {
+        if (--pending !== 0) return
+        self._closing = false
+        self._opened = false
+        cb()
+      }
+    }
+  }
+
+  ready (cb, waitForSource) {
+    const self = this
+    if (!this._opened) return this.open(() => this.ready(cb))
+
+    setImmediate(() => {
+      if (this.source.ready) this.source.ready(onsourceready)
+      else if (this._source.ready) this._source.ready(onsourceready)
+      else onsourceready()
+    })
+
+    function onsourceready () {
+      process.nextTick(() => {
+        if (self.status === Status.Ready) process.nextTick(cb)
+        else self.once('ready', cb)
+      })
+    }
+  }
+
+  pause () {
+    this.status = Status.Paused
+  }
+
+  resume () {
+    if (this.status !== Status.Paused) return
+    this.status = Status.Ready
+    if (!this._opened) return this.open()
+    this._run()
+  }
+
+  reset (cb = noop) {
+    const self = this
+    this.pause()
+    let pending = 1
+    process.nextTick(() => {
+      if (this._view.clearIndex) ++pending && this._view.clearIndex(done)
+      if (this._source.reset) ++pending && this._source.reset(done)
+      done()
+    })
+    function done () {
+      if (--pending !== 0) return
+      self.resume()
+      cb()
+    }
+  }
+
+  update () {
+    if (!this._opened) return
+    this.incomingUpdate = true
+    process.nextTick(this._run.bind(this))
+  }
+
+  _run () {
+    const self = this
+    if (!this._opened) return
+    if (this.status === Status.Running) return
+    if (this.status === Status.Paused) return
+
+    this.status = Status.Running
+
+    this._source.pull(onbatch)
+
+    function onbatch (result) {
+      // console.log(self.name, 'onbatch', result)
+      if (self.status === Status.Paused) return close()
+      if (!result) return close()
+      if (result.messages) result.messages = result.messages.filter(m => m)
+      if (!result.messages.length) return close(null, result)
+
+      const { messages = [], finished, onindexed } = result
+
+      // TODO: Handle timeout / error?
+      self._transform.run(messages, messages => {
+        if (!messages.length) return close(null, { messages, finished, onindexed })
+        self._view.map(messages, () => {
+          close(null, { messages, finished, onindexed })
+        })
+      })
+    }
+
+    function close (err, result) {
+      if (err) self.emit('error', err)
```

Comment:
Fixed this. However one thing I'm not sure about: If a view errors, should the flow be stopped / paused / run again? Currently, I think it would run in a loop which is not intended. Likely we want a new Error state for flows.